rfcs/2020-05-25-2685-dev-workflow-simplification.md

```diff
@@ -0,0 +1,244 @@
+
+```bash
+❯ cd /git/timberio/vector
+❯ direnv allow .
+❯ cd ../vector
+direnv: loading /git/timberio/vector/.envrc
+direnv: using nix
+cdirenv: export +AR +AS +CC +CONFIG_SHELL +CXX +HOST_PATH +IN_NIX_SHELL +LC_ALL +LD +NIX_BINTOOLS +NIX_BINTOOLS_WRAPPER_x86_64_unknown_linux_gnu_TARGET_HOST +NIX_BUILD_CORES +NIX_BUILD_TOP +NIX_CC +NIX_CC_WRAPPER_x86_64_unknown_linux_gnu_TARGET_HOST +NIX_CFLAGS_COMPILE +NIX_ENFORCE_NO_NATIVE +NIX_HARDENING_ENABLE +NIX_INDENT_MAKE +NIX_LDFLAGS +NIX_STORE +NM +OBJCOPY +OBJDUMP +PROTOC +PROTOC_INCLUDE +RANLIB +READELF +SIZE +SOURCE_DATE_EPOCH +STRINGS +STRIP +TEMP +TEMPDIR +TMP +TMPDIR +buildInputs +builder +configureFlags +depsBuildBuild +depsBuildBuildPropagated +depsBuildTarget +depsBuildTargetPropagated +depsHostHost +depsHostHostPropagated +depsTargetTarget +depsTargetTargetPropagated +doCheck +doInstallCheck +name +nativeBuildInputs +nobuildPhase +out +outputs +patches +phases +propagatedBuildInputs +propagatedNativeBuildInputs +shell +shellHook +stdenv +strictDeps +system ~LOCALE_ARCHIVE ~PATH
+```
+
+Now you can use the jobs detailed in **"Bring your own toolbox"** below. 
+
+### Bring your own toolbox
+
+> **Targets:** This option is required for MSVC/Mac/FreeBSD toolchains.
+
+To build Vector on your own host will require a fairly complete development environment!
+
+We keep an up to date list of all dependencies used in our CI environment inside our `default.nix` file. Loosely, you'll need the following:
+
+- **To build Vector:** Have working Rustup, C++/C build tools (LLVM, GCC, or MSVC), Python, and Perl, `make` (the GNU one preferably), `bash`, `cmake`, and `autotools`
+- **To run integration tests:** Have `docker` available, or a real live version of that service.
+- **To build the Website:** Have a working modern Ruby and Bundler toolchain available, also `bundle install` in the `scripts/` directory.
+- **To run the Website in Dev:** Have a working `node` environment with `npm`/`yarn`, also run `yarn` from the `website/` directory.
+- **To run `make check-component-features`:** Have `remarshal` installed.
+
+If you find yourself needing to run something (such as `make generate`) inside the Docker environment described above, that's totally fine, they won't collide or hurt each other. In this case, you'd just run `make environment-generate`.
+
+We're interested in reducing our dependencies if simple options exist. Got an idea? Try it out, we'd to hear of your successes and failures!
+
+In order to do your development on Vector, you'll primarily use a few commands, ordered from most to least frequently run:
+
+```bash
+# Validate your code can compile
+make check
+# Validate your code actually does compile (in dev mode)
+make build-dev
+# Validate your test pass
+make test scope="sources::example"
+# Validate tests (that do not require other services) pass
+make test
+# Validate your tests pass (starting required services in Docker)
+make test-integration scope="sources::example" autospawn=false
```

Comment:
Is `autospawn=false` a typo? The comment says it will start services.

I've also not seen this style of passing arguments to `make` before. I'm a little concerned about it being foreign both to users and to `make` itself.

Reply:
Really? I had no idea. WHat way is familiar? THis is just env vars?

Reply:
This is a pretty common idiom for setting (overriding) settings in makefiles on the command line. Mind you, the convention for settings is all-caps as indicated before, but I have seen both ways. For example, a common pattern for installing into an alternate path is `make install PREFIX=/tmp/fake/root`

Reply:
I'm fine with all caps :)